README

                           //\  //\  //\  //\  //\  //\  //\
                          //  \//  \//  \//  \//  \//  \//  \
                         //    _   _   _   _   _   _   _     \
                        //    / \ / \ / \ / \ / \ / \ / \     \
                       //    ( E ) X ) P ) E ) R ) M ) S )     \ 
                      //  __  \_/ \_/ \_/ \_/ \_/ \_/ \_/   __  \
                     // //  \                             //  \  \
                    // //    \  //\  //\  //\  //\  //\  //    \  \
                    \_//      \//  \//  \//  \//  \//  \//      \_//

                                   version 0.7 - 2013

                              Distributed under the GNU GPL
                         http://www.gnu.org/licenses/gpl-3.0.txt
                                Or see the file ./COPYING


==========
CONTENT:
==========
 - DESCRIPTION
 - DEPENDENCIES
 - USAGE
 - THE CONFIGURATION FILE
   - EXAMPLE
   - GENERAL SECTION
   - DEFAULT SECTION
   - DIRECTORY SECTION
 - RUNNING EXPERMS AS NON-ROOT USER
 - INOTIFY CONFIGURATION
 - FILES


========================================================================================
 ++ DESCRIPTION ++
========================================================================================
Runs as daemon and monitors file-changes happened in the directory set in experms.conf.
If changes happened, it adjusts the file-permissions and ownership/group.
You can either define one directory, or several sub-directories with different
ownerships and permissions.
It also allows exclusions based on directories or patterns (regex).
Further it is able to restore all the ownerships and permissions of all files based on
the config-file.
experms supports multiple instances, but only one per user.


========================================================================================
 ++ DEPENDENCIES ++
========================================================================================
experms depends on:

 - python2
 - python2-pyinotify (name can vary on different Linux distributions)
 - python2-psutil (name can vary on different Linux distributions)


========================================================================================
 ++ USAGE ++
========================================================================================
experms [ start | stop | restart | restore | status | log | err | dircount | 
(help|-h|--help) | version | foreground | debug [file] ]


      Explanation:

            start: Start experms
             stop: Stop experms
          restart: Restart experms
          restore: Restore the ownership and permissions of all files in the monitored
                   directory
           status: Displays status and if available PID of experms
              log: Display the log-entries in real-time
              err: Display the error-log-entries in real-time
         dircount: Displays the amount of directories being watched with actual
                   configuration. Useful to configure inotify
   help|-h|--help: Display a short help
          version: Displays the version of experms
       foreground: Start experms in the foreground. Useful to test a configuration
     debug [file]: Start experms in the foreground with much more output

NOTE: YOU NEED TO EDIT THE CONFIGURATION FILE BEFORE RUNNING EXPERMS THE FIRST TIME!


========================================================================================
 ++ THE CONFIGURATION FILE ++
========================================================================================
experms.conf is located under /etc/experms.conf. If you want to use experms as non-root
user (see 'man experms'), you need to make a copy of experms.conf in your
home-directory. It is also possible to hide the file (.experms.conf).

If changes happened to experms.conf, it is necessary to restart
experms: 'experms restart'.

=========
EXAMPLE:
=========
#
# /etc/experms.conf
#

[general]
log_activities = no
restore = no

[DEFAULT]
chmodd = 755

[directory_1]
path = /first/directory
user = first-user
group = first-group
chmodf = 640
chmodd = 750
excludepath = /first/directory/exclude1,/first/directory/subdir/exclude2

[directory_2]
path = /second/directory
group = second-group
chmodf = 0644
excludepattern = .txt|.TXT|.sh|.SH

=================
GENERAL SECTION:
=================
log_activities = yes | no
   Decide, if experms should write a log-file. If yes, the log will be
   stored in the directory, where experms lies. It also can be
   accessed with 'experms log'.

restore = yes | no
   Decide, if experms should restore all the ownerships and permissions
   of all files based on the config-file. In case there were changes
   while experms was not running.

=================
DEFAULT SECTION:
=================
Rules that should apply to all directory sections. They can be over‐
written inside a directory section.

===================
DIRECTORY SECTION:
===================
path = /path/to/watch/dir
   Set the directory where your rules should apply.

owner = username
   Set the owner of all the files and directories.

group = groupname
   Set the group of all the files and directories.

chmodf = octal permissions
   Set the permission of all the files.

chmodd = octal permissions
   Set the permission of all the directories.

excludepath = /path/to/exclude/dir,/path/to/exclude/file...
   Exclude some directories and files from the rules (comma  seperated).

excludepattern = regular expression
   Exclude files (and only files) based on a pattern.


========================================================================================
 ++ RUNNING EXPERMS AS NON-ROOT USER ++
========================================================================================
It's possible to run experms as non-root user. But there are quiet a few restictions:

 - Only root is able to change the owner of a file.
 - Only the owner of a file or root is able to change the group of a file.
 - Only the owner of a file or root is able to change the permissions of a file.

If you want to run experms as non-root user, make sure you are the owner of all
monitored files.

NOTE: If a file is modified by another user, it maybe changes it's ownership.
      In this case, experms would print a message to the error-log if executed as
      non\-root user.

If you want to use experms as non-root user (See 'man experms'), you need to make a copy
of experms.conf in your home-directory. It is also possible to hide the file
(.experms.conf).


========================================================================================
 ++ INOTIFY CONFIGURATION ++
========================================================================================
experms uses inotify to monitor the directories.
Inotify allows only a limited number of directories to watch per user.
Per default this is set to 8192.
You can increase this number by writing to /proc/sys/fs/inotify/max_user_watches
To make this change permanent, the following line should be added to /etc/sysctl.conf
or /etc/sysctl.d/99-sysctl.conf
fs.inotify.max_user_watches=8192 (your amount of directories)

You can check the amount of directories recursively with:

'experms dircount'


========================================================================================
 ++ FILES ++
========================================================================================
If executed as root, the scripts as well as the config-file should remain writable
only for root.

List of files:

./experms.py
/etc/experms.conf
./check_config.py
./daemon.py
/usr/share/man/man1/experms.1.gz
/usr/share/man/man5/experms.conf.5.gz
./COPYING
./README
./CHANGELOG

If experms was running at least once:

/var/log/experms.err
/var/log/experms.log

While Experms is running:

/tmp/experms_UID.pid



CHANGELOG
Experms changelog

Version 0.8
 - Tidying the code
 - Many bugfixes
 - 'experms status' will now exit with non-zero exit-code if it's not running (usefull to restart it with logrotate)

Version 0.7 - 2013-11-25
 - Support multiple instances (one per user)
 - The logs are written to /var/log/ if executed as root. If executed as normal user they lie in ~/
 - The config-file needs to be in /etc/ if executed as root. If executed as normal user it needs to be in ~/
 - Added 'debug' option
 - On every start the amount of watched directories will be displayed as well as the amount of allowed directories to monitor (inotify)
 - Errors during restore will be written to the error-log-file. A message will be printed.
 - Switch to psutil for checking if process is running
 - Added the possibility to display the version
 - Fixes in the documentation
 - Several Bugfixes
 - Cleaning up

Version 0.6 - 2013-08-22
 - Handle symlinks: Use owner- and group-rules for the link itself and not for the target-file. Do not chmod.
 - Make it possible to mixup the order of the 'general'- and 'directory'-sections.

Version 0.5 - 2013-08-21
 - Fixed a bug that prevented taking the actual owner and group if not definded in the config

Version 0.4 - 2013-08-21
 - Fixed a bug that prevented 'restore' and 'dircount' from working

Version 0.3 - 2013-08-21
- Rewrote the whole thing in Python. Nothing is the same as before ;-)
    Advantages:
    - doesn't crash ;-)
    - much faster
    - more lightweight
    - handle nested directories
    - exclude files and folders
    - exclude files based on a pattern (regex)
    - more intuitive configuration

Version 0.2 - 2013-01-26
- Fixed the annoying bug that prevented Experms from being started again after a reboot or crash


daemon.py
#!/usr/bin/env python2

import sys, os, time, atexit
from signal import SIGTERM

class Daemon:
        """
        A generic daemon class.

        Usage: subclass the Daemon class and override the run() method
        """
        def __init__(self, pidfile, stdin='/dev/null', stdout='/dev/null', stderr='/dev/null'):
                self.stdin = stdin
                self.stdout = stdout
                self.stderr = stderr
                self.pidfile = pidfile
        
        # If the .log or .err file is deleted, redirect standard file descriptors again
        def mknewlog(self):
                si = file(self.stdin, 'r')
                so = file(self.stdout, 'a+')
                se = file(self.stderr, 'a+', 0)
                os.dup2(si.fileno(), sys.stdin.fileno())
                os.dup2(so.fileno(), sys.stdout.fileno())
                os.dup2(se.fileno(), sys.stderr.fileno())
                
        def daemonize(self):
                """
                do the UNIX double-fork magic, see Stevens' "Advanced
                Programming in the UNIX Environment" for details (ISBN 0201563177)
                http://www.erlenstar.demon.co.uk/unix/faq_2.html#SEC16
                """
                try:
                        pid = os.fork()
                        if pid > 0:
                                # exit first parent
                                sys.exit(0)
                except OSError, e:
                        sys.stderr.write("fork #1 failed: %d (%s)\n" % (e.errno, e.strerror))
                        sys.exit(1)

                # decouple from parent environment
                os.chdir("/")
                os.setsid()
                os.umask(0)

                # do second fork
                try:
                        pid = os.fork()
                        if pid > 0:
                                # exit from second parent
                                sys.exit(0)
                except OSError, e:
                        sys.stderr.write("fork #2 failed: %d (%s)\n" % (e.errno, e.strerror))
                        sys.exit(1)

                # redirect standard file descriptors
                sys.stdout.flush()
                sys.stderr.flush()
                si = file(self.stdin, 'r')
                so = file(self.stdout, 'a+')
                se = file(self.stderr, 'a+', 0)
                os.dup2(si.fileno(), sys.stdin.fileno())
                os.dup2(so.fileno(), sys.stdout.fileno())
                os.dup2(se.fileno(), sys.stderr.fileno())

                # write pidfile
                atexit.register(self.delpid)
                pid = str(os.getpid())
                file(self.pidfile,'w+').write("%s\n" % pid)

        def delpid(self):
                os.remove(self.pidfile)

        def start(self):
                """
                Start the daemon
                """
                # Check for a pidfile to see if the daemon already runs
                try:
                        pf = file(self.pidfile,'r')
                        pid = int(pf.read().strip())
                        pf.close()
                except IOError:
                        pid = None
                        
                if pid:
                        message = "pidfile %s already exist. Experms already running?\n"
                        sys.stderr.write(message % self.pidfile)
                        sys.exit(1)

                # Start the daemon
                self.daemonize()
                self.run()

        def stop(self):
                """
                Stop the daemon
                """
                # Get the pid from the pidfile
                try:
                        pf = file(self.pidfile,'r')
                        pid = int(pf.read().strip())
                        pf.close()
                except IOError:
                        pid = None

                if not pid:
                        message = "pidfile %s does not exist. Experms not running?\n"
                        sys.stderr.write(message % self.pidfile)
                        return False# not an error in a restart

                # Try killing the daemon process       
                try:
                        while 1:
                                os.kill(pid, SIGTERM)
                                time.sleep(0.1)
                except OSError, err:
                        err = str(err)
                        if err.find("No such process") > 0:
                                if os.path.exists(self.pidfile):
                                        os.remove(self.pidfile)
                        else:
                                print str(err)
                                sys.exit(1)

        def restart(self):
                """
                Restart the daemon
                """
                self.stop()
                self.start()

        def run(self):
                """
                You should override this method when you subclass Daemon. It will be called after the process has been
                daemonized by start() or restart().
                """

daemon-functions.sh
#!/bin/bash

## ################################ ##
##                                  ##
## daemon-functions.sh VERSION 0.1a ##
## experms VERSION 0.2              ##
##                                  ##
## ################################ ##

## Distrubuted under the GPL
## http://www.gnu.org/licenses/gpl-3.0.txt

## BEWARE: This is a hardly modified version for Experms. You will find the original file here: http://blog.apokalyptik.com/2008/05/09/as-close-to-a-real-daemon-as-bash-scripts-get/

## No warranty of any kind... May run 
## off with your daughter. May explode
## in a ball of smoke and fire. Might
## work. Use at your own risk

## #!/bin/bash
##
## # Example Usage: datelogger.sh
## #    A sample daemon which simply logs the
## #    date and time once per second.
##
## function payload() {
##   while [ true ]; do
##     checkforterm
##     date
##     sleep 1
##   done
## }
##
## source /path/to/daemon-functions.sh

function daemonize() {
    echo $MY_PID > $MY_PIDFILE
    exec 3>&-           # close stdin
    exec 2>>$MY_ERRFILE # redirect stderr
    exec 1>>$MY_LOGFILE # redirect stdout
    echo $(date)" Daemonizing" >> $MY_ERRFILE
}

function checkforterm() {
    if [ -f $MY_KILLFILE ]; then
        echo $(date)" Terminating gracefully" >> $MY_ERRFILE
        rm $MY_PIDFILE
        rm $MY_KILLFILE
        inotifypid=$(cat /tmp/perms_pid.sfhsfhi)
        rm -f /tmp/experms.conf
        rm -f /tmp/perms_pid.sfhsfhi
        kill $inotifypid
        echo "done!"
        kill $MY_PID
        exit 0
    fi
}

MY_PID=$$
# MY_PATH=$(readlink -f $0)
# MY_ROOT=$(dirname $MY_PATH)
MY_NAME=$(basename $MY_PATH)
MY_PIDFILE="$MY_ROOT/.$MY_NAME.pid"
MY_KILLFILE="$MY_ROOT/.$MY_NAME.kill"
MY_ERRFILE="$MY_ROOT/.$MY_NAME.err"
MY_LOGFILE="$MY_ROOT/.$MY_NAME.log"
MY_WAITFILE="$MY_ROOT/.$MY_NAME.wait"
MY_BLOCKFILE="$MY_ROOT/.$MY_NAME.block"

CR="
"
SP=" "
OIFS=$IFS

case $1 in
#    pause)
#        touch $MY_WAITFILE
#        ;;
#    resume)
#        rm $MY_WAITFILE
#        ;;
    restart)
        $0 stop
        $0 start
        ;;
    start)
        if [ -f $MY_BLOCKFILE ]; then
            echo "Experms execution has been disabled"
            exit 0
        fi
        if [ -f $MY_PIDFILE ]; then
            if ps -Ao pid,fname | egrep "^$(cat $MY_PIDFILE) " | egrep -q "$MY_NAME"; then
                echo "Experms is already running"
                exit 0
            else
                rm $MY_PIDFILE
                rm -f /tmp/experms.conf
                rm -f /tmp/perms_pid.sfhsfhi
            fi
        fi
        if [ "$restore_at_start" == "yes" ]
        then
            func_restore
        fi
        $0 run &
        echo "Experms started"
        exec 3>&- # close stdin
        exec 2>&- # close stderr
        exec 1>&- # close stdout
        exit 0
        ;;
    disable)
        touch $MY_BLOCKFILE
        $0 stop
        ;;
    enable)
        if [ -f $MY_BLOCKFILE ]; then rm $MY_BLOCKFILE; fi
        ;;
    stop)
        echo -n "Terminating daemon... "
        $0 stat 1>/dev/null 2>/dev/null
        if [ $? -ne 0 ]; then
            echo "process is not running"
            exit 0
        fi
        touch $MY_KILLFILE
        checkforterm
        ;;
    stat)
        if [ -f $MY_BLOCKFILE ]; then
            echo "Experms execution disabled"
        fi
        if [ ! -f $MY_PIDFILE ]; then
            echo "$MY_NAME is not running"
            exit 1
        fi
        pgrep -l -f "$MY_NAME run" | grep -q -E "^$(cat $MY_PIDFILE) " 
        if [ $? -eq 0 ]; then
            echo "$MY_NAME is running with PID "`cat $MY_PIDFILE`
            if [ -f /tmp/perms_pid.sfhsfhi ]; then
                inotifypid=$(cat /tmp/perms_pid.sfhsfhi)
                echo "inotifywait is running with PID $inotifypid"
            fi
            exit 0
        else
            echo "$MY_NAME is not running"
            exit 1
        fi
        ;;
    log|stdout)
        if [ -f $MY_LOGFILE ] && [ -s "$MY_LOGFILE" ]; then
            tail -f $MY_LOGFILE
        else
            echo "No stdout output yet"
        fi
        ;;
    
    err|stderr)
        if [ -f $MY_ERRFILE ] && [ -s "$MY_ERRFILE" ]; then
            tail -f $MY_ERRFILE
        else
            echo "No stderr output yet"
        fi
        ;;
    run)
        daemonize
        payload
        ;;
    restore)
        if [ -f $MY_BLOCKFILE ]; then
            echo "Experms execution has been disabled"
            exit 0
        fi
        func_restore
        ;;
    help|?|--help|-h)
                echo """
Usage: $0 [ start | stop | restart | stat | disable | enable | (log|stdout) | (err|stderr) | restore | help ]

Experms runs as daemon and monitors file-changes happened in the directory set in experms.conf.
If changes happened, it adjusts the file-permissions and ownership/group.
You can either define one directory, or several sub-directories with different ownerships and permissions.
Further it is able to restore all the ownerships and permissions of all files based on the config-file.
"""
        exit 0
        ;;
    *)
        echo "Invalid argument"
        echo
        $0 help
        ;;
esac



check_config.py
#!/usr/bin/env python2
# -*- coding: utf-8 -*-

##
##         //\  //\  //\  //\  //\  //\  //\
##        //  \//  \//  \//  \//  \//  \//  \
##       //    _   _   _   _   _   _   _     \
##      //    / \ / \ / \ / \ / \ / \ / \     \
##     //    ( E ) X ) P ) E ) R ) M ) S )     \
##    //  __  \_/ \_/ \_/ \_/ \_/ \_/ \_/   __  \
##   // //  \                             //  \  \
##  // //    \  //\  //\  //\  //\  //\  //    \  \
##  \_//      \//  \//  \//  \//  \//  \//      \_//
##
##                 version 0.7 - 2013
##


##===========================================================================================================================
##
##         FILE: check-config.py
##
##  DESCRIPTION: Parses the configuration file, check validity and store content to variables
##
##       AUTHOR: Fabio Rämi - fabio(a)dynamix-tontechnik.ch
##
##      VERSION: 0.7
##
##      LICENCE: GNU GPL v3.0 or later.
##               http://www.gnu.org/licenses/gpl-3.0.txt
##               Experms comes with absolutely no warranty!
##               
##      COPYING: This file is part of Experms.
##
##               Experms is free software: you can redistribute it and/or modify
##               it under the terms of the GNU General Public License as published by
##               the Free Software Foundation, either version 3 of the License, or
##               (at your option) any later version.
##           
##               Experms is distributed in the hope that it will be useful,
##               but WITHOUT ANY WARRANTY; without even the implied warranty of
##               MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
##               GNU General Public License for more details.
##           
##               You should have received a copy of the GNU General Public License
##               along with Experms.  If not, see <http://www.gnu.org/licenses/>.
##
##      CREATED: 2013
##
##===========================================================================================================================

# bold: "\033[1m"
# green: "\033[32;1m"
# red: "\033[31;1m"
# normal: "\033[0m"

import sys
from os import path as os_path
from pwd import getpwnam, getpwuid
from grp import getgrnam, getgrgid
from re import compile as re_compile
from ConfigParser import SafeConfigParser,MissingSectionHeaderError


# function to verify the given octal permissions
def checkoctalperms(octalperms, chmodfd, i, debug):
  if chmodfd == 'chmodf':
    anrede = 'chmodf'
  else:
    anrede = 'chmodd'
  if octalperms != ['']:
    try:
      int(octalperms)
    except ValueError:
      print >> sys.stderr, "\033[31;1mError in section", i + ":", anrede, "is not an interger.\033[0m"
      return False
    if len(octalperms) != 3 and len(octalperms) != 4:
      print >> sys.stderr, "\033[31;1mError in section", i + ":", anrede, "needs to be a three or four digit octal number.\033[0m"
      return False
    for thing in octalperms:
      if int(thing) > 7:
        print >> sys.stderr, "\033[31;1mError in section", i + ":", anrede, "is not an octal number.\033[0m"
        return False
    if debug == True:
      print "'" + anrede + "' in section '" + i + "' is valid"
    return True

class Check(object):
  def __init__(self, uid, debug):
    # variable names from the configfile
    log_activities = 'log_activities'
    restore = 'restore'
    dirname = 'path'
    owner = 'owner'
    group = 'group'
    chmodf = 'chmodf'
    chmodd = 'chmodd'
    excludedir = 'excludepath'
    excludepattern = 'excludepattern'
    
    # default values for the section general
    logitdefault = 'no'
    restoredefault = 'no'
    
    # create the needed lists
    self.doit = []
    self.dirname = []
    self.owner = []
    self.group = []
    self.chmodf = []
    self.chmodd = []
    self.excludedir = []
    self.excludepattern = []
    
    #prepare the error variable
    errorsoccured = False
    
    # check for the existence of a config-file
    if uid == 0:
      if os_path.isfile('/etc/experms.conf'):
        configfile = '/etc/experms.conf'
      elif os_path.isfile(sys.path[0] + '/experms.conf'):
        configfile = sys.path[0] + '/experms.conf'
      else:
        print >> sys.stderr, "\033[31;1mError: No configuration-file (/etc/experms.conf) was found.\033[0m"
        sys.exit(1)
    else:
      home = os_path.expanduser("~")
      if os_path.isfile(home + '/.experms.conf'):
        configfile = home + '/.experms.conf'
      elif os_path.isfile(home + '/experms.conf'):
        configfile = home + '/experms.conf'
      elif os_path.isfile(sys.path[0] + '/experms.conf'):
        configfile = sys.path[0] + '/experms.conf'
      else:
        print >> sys.stderr, "\033[31;1mError: No configuration-file (~/experms.conf or ~/.experms.conf) was found.\033[0m"
        sys.exit(1)
    print "Using configuration-file '" + configfile + "'"
    
    # parse the config-file
    parser = SafeConfigParser()
    try:
      parser.read(configfile)
    except MissingSectionHeaderError:
      pass

    if parser.has_section('general'):
      if debug == True:
        print "section 'general' was found"
      if parser.has_option('general', log_activities):
        self.logit = parser.get('general', log_activities).lower()
        if self.logit == 'yes':
          if debug == True:
            print 'experms will write a log'
        elif self.logit == 'no' or self.logit == '':
          self.logit = logitdefault
          if debug == True:
            print "experms won't write a log"
        else:
          print >> sys.stderr, "\033[31;1mError: 'log_activities' must be either 'yes' or 'no'\033[0m"
          errorsoccured = True
      else:
        self.logit = logitdefault
        if debug == True:
          print "experms won't write a log"
      
      if parser.has_option('general', restore):
        self.restore = parser.get('general', restore).lower()
        if self.restore == 'yes':
          if debug == True:
            print 'experms will restore at start'
        elif self.restore == 'no' or self.restore == '':
          self.restore = restoredefault
          if debug == True:
            print "experms won't restore at start"
        else:
          print >> sys.stderr, "\033[31;1mError: 'restore' must be either 'yes' or 'no'\033[0m"
          errorsoccured = True
      else:
        self.restore = restoredefault
        if debug == True:
          print "experms won't restore at start"
    
      if len(parser.sections()) < 2:
        print >> sys.stderr, "\033[31;1mError: No directory-section was found.\nIf you have started experms for the first time, please edit the configfile first (usually /etc/experms.conf)\033[0m"
        errorsoccured = True
    else:
      self.restore = restoredefault
      self.logit = logitdefault
      if len(parser.sections()) < 1:
        print >> sys.stderr, "\033[31;1mError: No directory-section was found.\nIf you have started experms for the first time, please edit the configfile first (usually /etc/experms.conf)\033[0m"
        errorsoccured = True
    


    self.sectionname = []
    # can't use enumerate here, to make it possible to mixup the order between general and directory sections in the config
    number = -1
    for i in parser.sections():
      if i == 'general':
        continue
      number = number + 1
      self.sectionname.append(i)
      usowchmoderr = True
      self.doit.append('')
      
      if parser.has_option(i, dirname):
        self.dirname.append('')
        self.dirname[number] = parser.get(i, dirname).rstrip('/')
        if self.dirname[number] == '':
          print >> sys.stderr, "\033[31;1mError in section", i + ": 'path' is empty.\nIf you have started experms for the first time, please edit the configfile first.\033[0m"
          errorsoccured = True
        else:
          if not os_path.isdir(self.dirname[number]):
            print >> sys.stderr, "\033[31;1mError in section", i + ": 'path'", self.dirname[number], "doesn't exist\033[0m"
            errorsoccured = True
          else:
            if debug == True:
              print "'dirname' in section '" + i + "' is valid"
      else:
        print >> sys.stderr, "\033[31;1mError in section", i + ": 'dirname' is not set.\033[0m"
        errorsoccured = True
      
      self.owner.append('')  
      if parser.has_option(i, owner):
        self.owner[number] = parser.get(i, owner)
        if self.owner[number] != '':
          try:
            self.owner[number] = int(self.owner[number])
          except ValueError:
            try:
              getpwnam(self.owner[number])
            except KeyError:
              print >> sys.stderr, "\033[31;1mError in section", i + ": User", self.owner[number], "doesn't exist.\033[0m"
              errorsoccured = True
            else:
              # save the user as uid
              self.owner[number] = getpwnam(self.owner[number]).pw_uid
              usowchmoderr = False
              self.doit[number] = 1
              if debug == True:
                print "'user' in section '" + i + "' is valid"
          else:
            try:
              getpwuid(self.owner[number])
            except KeyError:
              print >> sys.stderr, "\033[31;1mError in section", i + ": User", self.owner[number], "doesn't exist.\033[0m"
              errorsoccured = True
            else:
              usowchmoderr = False
              self.doit[number] = 1
              if debug == True:
                print "'user' in section '" + i + "' is valid"
            
        else:
          self.owner[number] = -1
          self.doit[number] = 0
      else:
        self.owner[number] = -1
        self.doit[number] = 0
      
      self.group.append('')
      if parser.has_option(i, group):
        self.group[number] = parser.get(i, group)
        if self.group[number] != '':
          try:
            self.group[number] = int(self.group[number])
          except ValueError:
            try:
              getgrnam(self.group[number])
            except KeyError:
              print >> sys.stderr, "\033[31;1mError in section", i + ": Group", self.group[number], "doesn't exist.\033[0m"
              errorsoccured = True
            else:
              # save the group as gid
              self.group[number] = getgrnam(self.group[number]).gr_gid
              usowchmoderr = False
              self.doit[number] = 1
              if debug == True:
                print "'group' in section '" + i + "' is valid"
          else:
            try:
              getgrgid(self.group[number])
            except KeyError:
              print >> sys.stderr, "\033[31;1mError in section", i + ": Group", self.group[number], "doesn't exist.\033[0m"
              errorsoccured = True
            else:
              usowchmoderr = False
              self.doit[number] = 1
              if debug == True:
                print "'group' in section '" + i + "' is valid"
        else:
          self.group[number] = -1
      else:
        self.group[number] = -1
      
      self.chmodf.append('')
      if parser.has_option(i, chmodf):
        self.chmodf[number] = parser.get(i, chmodf)
        if self.chmodf[number] != '':
          if checkoctalperms(self.chmodf[number], 'chmodf', i, debug):
            if len(self.chmodf[number]) == 3:
              self.chmodf[number] = '0' + self.chmodf[number]
              self.chmodf[number] = int(self.chmodf[number], 8)
            elif len(self.chmodf[number]) == 4:
              self.chmodf[number] = int(self.chmodf[number], 8)
            usowchmoderr = False
            self.doit[number] = self.doit[number] + 2
          else:
            errorsoccured = True
      
      self.chmodd.append('')  
      if parser.has_option(i, chmodd):
        self.chmodd[number] = parser.get(i, chmodd)
        if self.chmodd[number] != '':
          if checkoctalperms(self.chmodd[number], 'chmodd', i, debug):
            if len(self.chmodd[number]) == 3:
              self.chmodd[number] = '0' + self.chmodd[number]
              self.chmodd[number] = int(self.chmodd[number], 8)
            elif len(self.chmodd[number]) == 4:
              self.chmodd[number] = int(self.chmodd[number], 8)
            usowchmoderr = False 
            self.doit[number] = self.doit[number] + 4 
          else:
            errorsoccured = True
      
      self.excludedir.append([])
      if parser.has_option(i, excludedir):
        exvalid = True
        self.excludedir[number] = parser.get(i, excludedir).split(',')
        for nr, item in enumerate(self.excludedir[number]):
          item = item.strip().rstrip('/')
          self.excludedir[number][nr] = item
          if item == '':
            self.excludedir[number].remove(item)
          else:
            if not os_path.isdir(item) and not os_path.isfile(item):
              print >> sys.stderr, "\033[31;1mError in section", i + ": 'excludedir'", item, "doesn't exist.\033[0m"
              errorsoccured = True
              exvalid = False
        if exvalid == True:
          if self.dirname[number] in self.excludedir[number]:
            print >> sys.stderr, "\033[31;1mError in section", i + ": 'excludedir'", item, "is the same like 'dirname'.\033[0m"
            errorsoccured = True
          if debug == True:
            print "'excludedir' in section '" + i + "' is valid"
        if self.excludedir[number] == [] or self.excludedir[number] == ['']:
          self.excludedir[number] = None
      else:
        self.excludedir[number] = None
      
      self.excludepattern.append('')
      if parser.has_option(i, excludepattern):
        exvalid = True
        self.excludepattern[number] = parser.get(i, excludepattern)
        try:
          re_compile(self.excludepattern[number])
        except:
          print >> sys.stderr, "\033[31;1mError in section", i + ": 'excludepattern' must be a regular expression.\033[0m"
          errorsoccured = True
        else:
          if debug == True:
            print "'excludepattern' in section '" + i + "' is valid"
        if self.excludepattern[number] == '':
          self.excludepattern[number] = None
      else:
        self.excludepattern[number] = None
      
      if usowchmoderr == True:
        print >> sys.stderr, "\033[31;1mError in section", i + ": With your actual configuration, experms will do exactly nothing.\033[0m"
        errorsoccured = True
    
    if errorsoccured == True:
      print >> sys.stderr, "\033[31;1mAborting!\033[0m"
      sys.exit(1)